java/vector/src/main/java/org/apache/arrow/vector/stream/MessageSerializer.java

```diff
@@ -78,12 +78,25 @@ public static int bytesToInt(byte[] bytes) {
    * @throws IOException if something went wrong
    */
   public static long serialize(WriteChannel out, Schema schema) throws IOException {
+    long start = out.getCurrentPosition();
+    assert start % 8 == 0;
+
     FlatBufferBuilder builder = new FlatBufferBuilder();
     int schemaOffset = schema.getSchema(builder);
     ByteBuffer serializedMessage = serializeMessage(builder, MessageHeader.Schema, schemaOffset, 0);
-    long size = out.writeIntLittleEndian(serializedMessage.remaining());
-    size += out.write(serializedMessage);
-    return size;
+
+    int size = serializedMessage.remaining();
+    // ensure that message aligns to 8 byte padding - 4 bytes for size, then message body
+    if ((size + 4) % 8 != 0) {
```

Comment:
I think the padding code can use some clean up and refactor, but we can merge this first and clean up as follow up.